lib/web/desktop.go

```diff
@@ -103,17 +102,36 @@ func createDesktopConnection(
 	//
 	// In the future, we may want to do something smarter like latency-based
 	// routing.
-	service := winServices[rand.Intn(len(winServices))]
-	log = log.WithField("windows-service-uuid", service.GetName())
-	log = log.WithField("windows-service-addr", service.GetAddr())
-	serviceCon, err := site.DialTCP(reversetunnel.DialParams{
-		From:     &utils.NetAddr{AddrNetwork: "tcp", Addr: r.RemoteAddr},
-		To:       &utils.NetAddr{AddrNetwork: "tcp", Addr: service.GetAddr()},
-		ConnType: types.WindowsDesktopTunnel,
-		ServerID: service.GetName() + "." + ctx.parent.clusterName,
-	})
+	winDesktops, _ := ctx.unsafeCachedAuthClient.GetWindowsDesktops(r.Context())
+	var validServices []types.WindowsDesktopService
+	for _, desktop := range winDesktops {
+		for _, service := range winServices {
+			if desktop.GetHostID() == service.GetName() {
+				validServices = append(validServices, service)
+			}
+		}
+	}
+
+	var serviceCon net.Conn
+	for _, service := range validServices {
+		log = log.WithField("windows-service-uuid", service.GetName())
+		log = log.WithField("windows-service-addr", service.GetAddr())
+		serviceCon, err = site.DialTCP(reversetunnel.DialParams{
+			From:     &utils.NetAddr{AddrNetwork: "tcp", Addr: r.RemoteAddr},
+			To:       &utils.NetAddr{AddrNetwork: "tcp", Addr: service.GetAddr()},
+			ConnType: types.WindowsDesktopTunnel,
+			ServerID: service.GetName() + "." + ctx.parent.clusterName,
+		})
+		if err != nil {
```

Comment:
Don't forget, if err is nil then you connected successfully and you don't need to retry, you can break out of the for loop.